LG-278: handle stack traces written to System.out / System.err as a single event in logs

Channyboy opened the pull request:
# ** Squashed! **

Epic:
https://github.com/OpenLiberty/open-liberty/issues/1598

Related issue:
#13681 

Migrated from:
https://github.com/OpenLiberty/open-liberty/pull/16276

- Have only additionally completed the FAT tests

Work previously conducted by @kabicin  @adriankalafut @fwji 

#build

skasund commented:
L2 message review not required

Channyboy commented:
#build
#libby

Channyboy commented:
Single failure in an unrelated component, re-spinning.
#build

Channyboy commented:
Build failures are different from previous build and are unrelated to the changes of this PR.
Related RTC defect to failure seen which has been cancelled/returned due to intermittent occurrence.

Channyboy commented:
Commits after@adriankalafut's last commit at `171843fe352054d53f57b7384b180c016007f408` are new

Channyboy commented:
#build
#libby

Channyboy commented:
Latest build contains errors/failures in unrelated components.

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/Activator.java:
```diff
@@ -17,30 +17,72 @@
+			this.proxyActivator.activate();
+		}
+	
+	}
```
indentation

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/Activator.java:
```diff
@@ -17,30 +17,72 @@
+		if (ProductInfo.getBetaEdition()) {	
+			this.proxyActivator.deactivate();
+		}
+    }
```
indentation

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/stackjoiner/ThrowableProxyActivator.java:
```diff
@@ -0,0 +1,432 @@
+        }
+        
+        // Otherwise, check if WLP_LOGGING_STACK_JOIN has been set in the server.env
+        // Or if com.ibm.ws.logging.stack.join has been set in the bootstrap.properties 
```
indentation

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/stackjoiner/ThrowableProxyActivator.java:
```diff
@@ -0,0 +1,432 @@
+			stackJoinConfig =  (Boolean) serverXmlHashMap.get("stackJoin");
+		}
+        if(stackJoinConfig != null && stackJoinConfig == true) {
+        	return Boolean.TRUE;
```
just use `true` here

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/stackjoiner/ThrowableProxyActivator.java:
```diff
@@ -0,0 +1,432 @@
+        // Or if com.ibm.ws.logging.stack.join has been set in the bootstrap.properties 
+		boolean isEnabled = StackJoinerConfigurations.getInstance().stackJoinerEnabled();
+		if(isEnabled) {
+			return Boolean.TRUE;
```
just use `true` here

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/stackjoiner/ThrowableProxyActivator.java:
```diff
@@ -0,0 +1,432 @@
+			return Boolean.TRUE;
+		}
+		
+		return Boolean.FALSE;
```
just use `false` here

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/stackjoiner/ThrowableProxyActivator.java:
```diff
@@ -0,0 +1,432 @@
+	private static volatile boolean throwableProxyIsActivated = false;
+	/**
+	 * The target package for the boot loader delegated classes.
+	 */
```
squash comments like this (throughout the file to one line):
```
/** The target package for the boot loader delegated classes. */
```

please also fix indent of this whole file to our usual 4 space indents

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/stackjoiner/MethodProxy.java:
```diff
@@ -0,0 +1,68 @@
+	}
+	
+	@SuppressWarnings("rawtypes")
+	private Class<?> retrieveClass(Instrumentation inst, String classGroup) {
```
not sure where param name came from for this... probably should just be className rather than classGroup

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/stackjoiner/ThrowableProxyActivator.java:
```diff
@@ -0,0 +1,432 @@
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.Bundle;
+
+public class ThrowableProxyActivator {
```
suggest you rename this class to StackJoinerActivator

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/stackjoiner/ThrowableProxyActivatorStorage.java:
```diff
@@ -0,0 +1,43 @@
+import java.lang.instrument.Instrumentation;
+import org.osgi.framework.BundleContext;
+
+public class ThrowableProxyActivatorStorage{
```
should make be an inner class of ThrowableProxyActivator

Channyboy replied:
Class removed, values are now set statically inside `StackJoinerActivator`

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/stackjoiner/ThrowableProxyActivator.java:
```diff
@@ -0,0 +1,432 @@
+
+			// Don't proceed with ThrowableProxy activation if the feature has already been
+			// activated
+			if (throwableProxyIsActivated) {
```
rename to throwableClassIsTransformed

Channyboy replied:
Used  `isThrowableClassTransformed` instead

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/stackjoiner/ThrowableProxyActivator.java:
```diff
@@ -0,0 +1,432 @@
+	 * Returns true if the stack joiner feature has been enabled, otherwise false
+	 * @return true if the stack joiner feature has been enabled, otherwise false
+	 */
+	public static boolean isEnabled() {
```
rename to isEnabledInConfig

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/stackjoiner/ThrowableProxyActivator.java:
```diff
@@ -0,0 +1,432 @@
+				methodProxy = new MethodProxy(inst, BASE_TRACE_SERVICE_CLASS_NAME, BASE_TRACE_SERVICE_METHOD_NAME,
+						Throwable.class, PrintStream.class);
+			}
+			if (methodProxy != null && methodProxy.isInitialized()) {
```
simplify this (slightly) by adding a check after line 125 to see if methodProxy is initialized, and if not, set methodProxy to null...
```
if (!methodProxy.isInitialized()) 
    methodProxy = null;
```

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/stackjoiner/ThrowableProxyActivator.java:
```diff
@@ -0,0 +1,432 @@
+	 *
+	 * @throws Exception the method invocation exception
+	 */
+	void activateThrowableProxyMethodTarget() throws Exception {
```
rename to bindThrowableProxyMethodTarget

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/stackjoiner/boot/templates/ThrowableProxy.java:
```diff
@@ -0,0 +1,57 @@
+import java.lang.reflect.Method;
+
+/**
+ * ThrowableProxy exposes an annotated BaseTraceService method to the Java Bootstrap ClassLoader
```
I don't think the BaseTraceService method is annotated.  Should comment be something like...
```
ThrowableProxy is used by code running in the bootstrap classloader to access a method in BaseTraceService
```

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/LoggingConfigurationService.java:
```diff
@@ -120,22 +125,32 @@ public synchronized void updated(Dictionary properties) throws ConfigurationExce
-        Map<String, Object> newMap = null;
         if (properties instanceof Map) {
-            newMap = (Map<String, Object>) properties;
+        	serverXmlMap = (Map<String, Object>) properties;
```
indentation is messed up on this line and many others in this file

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/LoggingConfigurationService.java:
```diff
@@ -21,13 +21,10 @@
 import java.util.Dictionary;
 import java.util.Enumeration;
 import java.util.HashMap;
-import java.util.HashSet;
```
need to update copyright year in this file

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/LoggingConfigurationService.java:
```diff
@@ -120,22 +125,32 @@ public synchronized void updated(Dictionary properties) throws ConfigurationExce
+			boolean stackJoinValue = false;
+			if (serverXmlMap.get(STACK_JOIN_SERVER_XML_CONFIG_NAME) != null
+					&& ((Boolean) serverXmlMap.get(STACK_JOIN_SERVER_XML_CONFIG_NAME)) == true) {
+				stackJoinValue = true;
```
rather than do the stack join enabled check here, we should:
- call ThrowableProxyActivator.isEnabled.  We could actually pass that method the serverXmlMap/newMap so that isEnabled doesn't need to call back to LoggingConfigurationService.getServerXmlHashMap()
- we could then get rid of getServerXmlHashMap

Channyboy replied:
code refactored out

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/LoggingConfigurationService.java:
```diff
@@ -266,4 +281,34 @@ private LogLevel mapLogLevel(String logLevelSpec) {
+        		return;
+        	}
+        	
+        	ThrowableProxyActivator throwableProxyActivator = new ThrowableProxyActivator(throwableProxyActivatorStorage.getInstrumentation(), throwableProxyActivatorStorage.getBundleContext());
```
add a getInstance singleton method to ThrowableProxyActivator and call that instead of creating a new instance of it here

Channyboy replied:
code refactored out

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/LoggingConfigurationService.java:
```diff
@@ -266,4 +281,34 @@ private LogLevel mapLogLevel(String logLevelSpec) {
+	        // if the feature should be deactivated
+	        else if(throwableProxyActivator != null) {
+	        	throwableProxyActivator.deactivate();
+	        }
```
remove "throwableProxyActivator != null" checks.. it can't be null here

Channyboy replied:
code refactored out

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/stackjoiner/ThrowableProxyActivator.java:
```diff
@@ -0,0 +1,432 @@
+			if (throwableProxyIsActivated) {
+				return;
+			}
+			if (isEnabled()) {
```
move the call to the isEnabled() check to the logging activator.  remove the call to isEnabled from activate.  calling activate will then mean you want to enable stackjoiner, and calling deactivate will mean you want to disable it.

Channyboy replied:
code refactored out

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/LoggingConfigurationService.java:
```diff
@@ -53,9 +53,13 @@
     /** reference to registered RAS config service */
     private ServiceRegistration<ManagedService> configRef = null;
+    
+    private static Map<String, Object> serverXmlMap = null;
```
can eliminate serverXmlMap (see other comments about this class)

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/LoggingConfigurationService.java:
```diff
@@ -266,4 +281,34 @@ private LogLevel mapLogLevel(String logLevelSpec) {
         return ht;
     }
+    
+    public static Map<String, Object> getServerXmlHashMap(){
```
remove this method (see other comments in this class)

donbourne commented on dev/com.ibm.ws.logging/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -83,7 +83,11 @@
+            
+		<AD name="%stackJoin" description="%stackJoin.desc"
+        	ibm:variable="com.ibm.ws.logging.stackJoin"
+            id="stackJoin" required="false" type="Boolean" default="false" ibm:beta="true" />
```
indentation

donbourne commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/LogProviderConfigImpl.java:
```diff
@@ -403,6 +410,10 @@ public String getMessageFormat() {
         return messageFormat;
     }
 
+    public boolean getStackJoinConfiguration() {
```
rename isStackJoin

Channyboy replied:
Should we also consider refactoring the class to update getter methods for other boolean values? (in another issue)
e.g. copySystemSTreams, AppsWriteJson, newLogsonStart ,etc

donbourne replied:
I don't think we need to... it was just a particularly misleading name to have a method called getStackJoinConfiguration returning a boolean.

donbourne commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/LogProviderConfigImpl.java:
```diff
@@ -96,6 +96,8 @@
     /** Format to use for messages.log */
     protected volatile String messageFormat = LoggingConstants.DEFAULT_MESSAGE_FORMAT;
 
+    protected volatile boolean stackJoinerConfiguration = false;
```
rename isStackJoinEnabled

Channyboy replied:
Used `stackJoin` instead so that the getter `isStackJoin()`

donbourne commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java:
```diff
@@ -242,6 +243,9 @@ protected StackTraceFlags initialValue() {
     };
 
+    public final static int BYTE_ARRAY_OUTPUT_BUFFER_THRESHOLD = ThreadLocalByteArrayOutputStream.getByteArrayOutputThreshold();
+    public volatile static boolean STACK_JOINER_FEATURE_ENABLED = false;
```
this isn't a constant, so shouldn't use block capitalization

donbourne commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/utils/StackJoinerConfigurations.java:
```diff
@@ -0,0 +1,33 @@
+
+import com.ibm.ws.logging.internal.impl.BaseTraceService;
+
+public class StackJoinerConfigurations {
```
not clear why we need this whole class just to return BaseTraceService.STACK_JOINER_FEATURE_ENABLED.  can the user of this class just call BaseTraceService.STACK_JOINER_FEATURE_ENABLED directly?

Channyboy replied:
The server start values are evaluated by the `com.ibm.ws.logging.osgi` bundle.
`BaseTraceService` is in an internal package in `com.ibm.ws.logging` and  the `com.ibm.ws.logging.osgi` bundle does not import any packages containing `internal`.

This is a workaround to that restriction.

donbourne commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java:
```diff
@@ -1567,7 +1576,10 @@ public synchronized void println(char c[]) {
             try {
-                super.print(s);
+                if (TrOutputStream.isPrintingStackTrace.get())
+                    super.println(s);
```
I believe our code in filterStackTraces is stripping the NLs off of the last line of the stack trace, which is necessitating this unusual need for println.  would be better if we could avoid stripping that NL in the first place so that this isn't needed.

Channyboy replied:
`filterStackTraces` already seems to be donig it.
Changing it back to `print()` doesn't seem to take any effect.

donbourne commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java:
```diff
@@ -1687,6 +1765,46 @@ protected synchronized void writeStreamOutput(SystemLogHolder holder, String txt
+     * @param originalStream reference to the PrintStream object to be written to
+     * @return true if the printStackTrace method was overridden, false otherwise
+     */
+    public static boolean printStackTraceOverride(Throwable t, PrintStream originalStream) {
```
rename to prePrintStackTrace since it's not really overriding, it's just happening before printStackTrace

donbourne commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java:
```diff
@@ -1687,6 +1765,46 @@ protected synchronized void writeStreamOutput(SystemLogHolder holder, String txt
     }
 
+    /**
+     * This method is accessed via reflection by the com.ibm.ws.logging.internal.osgi.stackjoiner.MethodProxy class
+     * It will be called on any Throwable.printStackTrace(PrintStream) invocation.
```
I'd suggest the following comment...

This method is accessed by bytecode injected into the start of the Throwable.printStackTrace(PrintStream) method.  When this method returns true, printStackTrace will not run the normal printStackTrace method body.  When this method returns false, printStackTrace will run the normal printStackTrace method body.

donbourne reviewed (changes requested):
I have only reviewed the runtime code (not the test code).

Channyboy commented:
Refactored configuration logic on how stack joiner feature is activated or disabled

    - Modified/reverted com.ibm.ws.logging.osgi bnd to original
    - Modified/reverted Activator class back to original with static fields for Instrumentation and BundleContext w/ accessors
    - Delete ThrowableProxyActivatorStorage, instrumentation and BundleContext are stored on a "bundle" level with the Activator and the StackJoinerActivator itself.
    - Renamed ThrowableProxyActivator to StackJoinerManager. Singleton set a static reference in LoggingCnfigurationService. It now has inits to retrieve instrumentation and bundlecontext from bundle Activator class. It is called each time updated() is called in LoggingConfigurationService to manage activation or deactivation of StackJoiner.
    - Modified LoggingConfigurationService to solely handle configuration of StackJoiner feature.
    - Change TeePrintStream's println impl to call super.print()
    
    
  #build #libby

donbourne commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java:
```diff
@@ -1617,7 +1629,10 @@ public synchronized void println(String s) {
+                if (TrOutputStream.isPrintingStackTrace.get())
+                    super.print(obj);
+                else
+                    super.print(obj);
```
I think you just need...
```
super.print(obj);
```

Channyboy replied:
fixed

donbourne commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java:
```diff
@@ -1606,7 +1615,10 @@ public synchronized void println(char c[]) {
+                if (TrOutputStream.isPrintingStackTrace.get())
+                    super.print(s);
+                else
+                    super.print(s);
```
I think you just need...
```
super.print(s);
```

Channyboy replied:
fixed

donbourne commented on dev/com.ibm.ws.logging.osgi/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -83,7 +83,11 @@
+		
+		<AD name="%stackJoin" description="%stackJoin.desc"
+        	ibm:variable="com.ibm.ws.logging.stackJoin"
+            id="stackJoin" required="false" type="Boolean" default="false" ibm:beta="true"/>
```
indentation

Channyboy replied:
fixed

donbourne commented on dev/com.ibm.ws.logging.osgi/src/com/ibm/ws/logging/internal/osgi/stackjoiner/StackJoinerManager.java:
```diff
@@ -0,0 +1,465 @@
+         * isStackJoinEnabled to false as a default with an "else".
+         * DON"T CHANGE TO THIS:
+         * 
+         * else { isStackJoinEnabled = false; }
```
great comment :-)

donbourne commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/utils/StackJoinerConfigurations.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    public boolean stackJoinerEnabled() {
+        return BaseTraceService.isStackJoinEnabled;
```
I think you can replace this class with...

```java
public class StackJoinerConfigurations {
    public static boolean stackJoinerEnabled() { return BaseTraceService.isStackJoinEnabled; }
}
```

Channyboy replied:
fixed

donbourne reviewed (changes requested):
I still haven't reviewed the test cases...

This is looking really cleaned up!  Just a few minor comments.

Channyboy commented:
#build #libby

donbourne commented:
LGTM, but I still need to review the test code

donbourne commented on dev/com.ibm.ws.logging_fat/publish/files/stackJoiner/server-disableStackJoiner.xml:
```diff
@@ -0,0 +1,22 @@
+
+    <include location="../fatTestPorts.xml"/>
+
+	<logging stackJoin="false" />
```
indentation

donbourne commented on dev/com.ibm.ws.logging_fat/publish/servers/com.ibm.ws.logging.stackjoiner.serverXMLConfig/server.xml:
```diff
@@ -0,0 +1,22 @@
+
+    <include location="../fatTestPorts.xml"/>
+
+	<logging stackJoin="true" />
```
indentation

donbourne reviewed (approved):
LGTM -- just a couple of minor formatting things left.

Channyboy commented:
#build #libby

Channyboy committed:
Single event stack traces:

Allow configuration to consolidate stack traces to be emitted as a single message over multiple messages (for each line of a stacktrace)

Configuration:
Bootstrap.properties > com.ibm.ws.logging.stackJoin=/truefalse
Server.env > WLP_LOGGING_STACK_JOIN=true/false
Server.xml >  <logging stackJoin="true/false" />

Uses ByteCode injection to toggle flag that allows for full stack to be printed in a single line.
Use of Instrumentation API to transform logging code.

Squashed commits include work from @kabicin @adriankalafut @fwji

Channyboy commented:
#build #libby

Channyboy commented:
#build #libby

Channyboy commented:
Error/failures encountered were in an unrelated component (GUI related). Regardless running another #build

Channyboy commented:
Errors/Failures are once again in non-related components. (GUI related).
